### PR TITLE
Enhance got command

### PR DESCRIPTION
**Pull request by lebr0nli**
In this PR, I wrote some enhancements for the `got` command which were inspired by bata24's great work in https://github.com/bata24/gef .

----

Summary of the enhancements:

- The new usage of the `got` command:
```
usage: got [-h] [-p PATH_FILTER | -a] [-r] [symbol_filter]

Show the state of the Global Offset Table.

Examples:
  got
  got puts
  got -p libc
  got -a

positional arguments:
 symbol_filter     Filter results by passed symbol name. (default: )

optional arguments:
 -h, --help      show this help message and exit
 -p PATH_FILTER, --path PATH_FILTER
            Filter results by passed library path. (default: )
 -a, --all       Parse all libraries including the target executable. (default: False)
 -r, --readonly    Set the flag to not filter out read-only entries. (default: False)
```

- We can show loaded libraries entries by something like:
screenshot-2
screenshot-4

- Before, we only show the relocation whose `r_info_type` is `*_JUMP_SLOT`, but now we also show the relocation whose type is `*_IRELATIVE` or `*_GLOB_DAT`. For example, now we have the ability to show some `R_X86_64_IRELATIVE` relocation in libc (e.g. `strlen` used by libc's `puts`):
screenshot-1

- We can filter out or not filter out the read-only entries (default will filter out read-only entries) by:
screenshot-3

---

The enhancements in this PR still have some todos and bugs in some cases(I wrote them in the comment), but this should at least resolve the requirements in #1734 for the ability to display libc's or any other loaded libraries's entries. (Currently, I didn't have a good idea to handle those todos and bugs properly, maybe we can fix them in a new PR if they don’t affect too much :p)

### CODE REVIEW

File: pwndbg/commands/got.py [hunk 1]
```diff
@@ -1,54 +1,171 @@
...
+    help="Parse all libraries including the target executable.",
+    action="store_true",
+    default=False,
+    dest="all_",
+)
 parser.add_argument(
-    "name_filter", help="Filter results by passed name.", type=str, nargs="?", default=""
+    "-r",
+    "--readonly",
+    help="Set the flag to not filter out read-only entries.",
```
Comment (disconnect3d) [inline pwndbg/commands/got.py]: Hmm... `--readonly` sounds as if we want only read-only entries...

Comment (disconnect3d) [reply pwndbg/commands/got.py]: Maybe `--show-ro` or `--show-readonly`?

File: pwndbg/wrappers/readelf.py [hunk 2]
```diff
@@ -1,29 +1,31 @@
...
-def get_jmpslots():
-    local_path = pwndbg.gdblib.file.get_proc_exe_file()
-    cmd = get_jmpslots.cmd + ["--relocs", local_path]
-    readelf_out = pwndbg.wrappers.call_cmd(cmd)
-
-    return filter(_extract_jumps, readelf_out.splitlines())
+class RelocationType(Enum):
+    JUMP_SLOT = 1  # e.g.: R_X86_64_JUMP_SLOT
+    GLOB_DAT = 2  # e.g.: R_X86_64_IRELATIVE
+    IRELATIVE = 3  # e.g.: R_X86_64_GLOB_DAT
```
Comment (disconnect3d) [inline pwndbg/wrappers/readelf.py]: Are the comments here correct? GLOB_DAT comment is on the IRELATIVE line etc.

Comment (disconnect3d) [reply pwndbg/wrappers/readelf.py]: Any source for those flags? Can we link it?

**Comment by disconnect3d** [comment]
For what is worth, while displaying GOT makes lots of sense, I usually do that with pure `telescope <rw page>`. Also, GOT is not the only thing that contains useful pointers, so ideally you would want to display "all possible code pointers" or hook/break on them somehow...

But anyway, this looks good; just needs small enhancements, see comments

File: pwndbg/commands/got.py [hunk 3]
```diff
@@ -1,54 +1,171 @@
...
+    got puts
+    got -p libc
+    got -a
+""",
+)
+group = parser.add_mutually_exclusive_group()
+group.add_argument(
+    "-p",
+    "--path",
+    help="Filter results by passed library path.",
```
Comment (disconnect3d) [inline pwndbg/commands/got.py]:
```suggestion
    help="Filter results by library/objfile path.",
```

File: pwndbg/commands/got.py [hunk 4]
```diff
@@ -1,54 +1,171 @@
...
+    type=str,
+    default="",
+    dest="path_filter",
+)
+group.add_argument(
+    "-a",
+    "--all",
+    help="Parse all libraries including the target executable.",
```
Comment (disconnect3d) [inline pwndbg/commands/got.py]:
```suggestion
    help="Process all libs/obfjiles including the target executable.",
```

File: pwndbg/commands/got.py [hunk 5]
```diff
@@ -1,54 +1,171 @@
...
-    "name_filter", help="Filter results by passed name.", type=str, nargs="?", default=""
+    "-r",
+    "--readonly",
+    help="Set the flag to not filter out read-only entries.",
+    action="store_true",
+    default=False,
+    dest="accept_readonly",
+)
+parser.add_argument(
+    "symbol_filter", help="Filter results by passed symbol name.", type=str, nargs="?", default=""
```
Comment (disconnect3d) [inline pwndbg/commands/got.py]:
```suggestion
    "symbol_filter", help="Filter results by symbol name.", type=str, nargs="?", default=""
```

File: pwndbg/commands/got.py [hunk 6]
```diff
@@ -1,54 +1,171 @@
...
-def got(name_filter="") -> None:
-    relro_status = pwndbg.wrappers.checksec.relro_status()
-    pie_status = pwndbg.wrappers.checksec.pie_status()
-    jmpslots = list(pwndbg.wrappers.readelf.get_jmpslots())
-    if not jmpslots:
-        print(message.error("NO JUMP_SLOT entries available in the GOT"))
+def got(path_filter, all_, accept_readonly, symbol_filter) -> None:
+    # Show the filters we are using
+    if path_filter:
+        print("Filtering by path keyword: " + message.hint(path_filter))
```
Comment (disconnect3d) [inline pwndbg/commands/got.py]:
```suggestion
        print("Filtering by lib/objfile path: " + message.hint(path_filter))
```

File: pwndbg/commands/got.py [hunk 7]
```diff
@@ -1,54 +1,171 @@
...
+    if symbol_filter:
+        print("Filtering by symbol keyword: " + message.hint(symbol_filter))
```
Comment (disconnect3d) [inline pwndbg/commands/got.py]:
```suggestion
        print("Filtering by symbol name: " + message.hint(symbol_filter))
```

File: pwndbg/commands/got.py [hunk 8]
```diff
@@ -1,54 +1,171 @@
...
+    if not accept_readonly:
+        print("Filtering out read-only entries.")
```
Comment (disconnect3d) [inline pwndbg/commands/got.py]:
```suggestion
        print("Filtering out read-only entries (display them with -r or --show-readonly)")
```

File: pwndbg/commands/got.py [hunk 9]
```diff
@@ -1,54 +1,171 @@
...
+    "--all",
+    help="Parse all libraries including the target executable.",
+    action="store_true",
+    default=False,
+    dest="all_",
+)
 parser.add_argument(
-    "name_filter", help="Filter results by passed name.", type=str, nargs="?", default=""
+    "-r",
+    "--readonly",
```
Comment (disconnect3d) [inline pwndbg/commands/got.py]:
```suggestion
    "--show-readonly",
```

File: pwndbg/commands/got.py (hunk 1 above)
Comment (disconnect3d) [inline pwndbg/commands/got.py]:
```suggestion
    help="Also display read-only entries (which are filtered out by default).",
```

Comment (disconnect3d) [reply pwndbg/commands/got.py]: Changed to -r --show-readonly.

**Comment by disconnect3d** [comment]
>  (Currently, I didn't have a good idea to handle those todos and bugs properly, maybe we can fix them in a new PR if they don’t affect too much :p)

Can we somehow inform the user that we may be missing some information?

I am going to go ahead and merge this PR if it passes CI. Please open another PR with further enhancements :)

File: pwndbg/wrappers/readelf.py (hunk 2 above)
Comment (lebr0nli) [reply pwndbg/wrappers/readelf.py]: > Are the comments here correct? GLOB_DAT comment is on the IRELATIVE line etc.

Ahh, I messed up the copy-pasting lol. Thanks, I'll fix it.

> Any source for those flags? Can we link it?

Sure, I'll add some links for those example flags.

**Comment by lebr0nli** [comment]
> so ideally you would want to display "all possible code pointers" or hook/break on them somehow...

Yeah, I think having the option to somehow hook/break on those functions might be more helpful than only displaying them, I’ll try to add this kind of feature in the future.

> Can we somehow inform the user that we may be missing some information?

Sure, I’ll add some hints to remind the users.

**Comment by disconnect3d** [comment]
The problem is you rather want to hook on call/jmp to code pointers and
since this may be realized in multiple ways I dont know how we could do sth
reliable here.

**Comment by disconnect3d** [comment]
@lebr0nli can you update tests after my changes so that we can merge this?

**Comment by lebr0nli** [comment]
> can you update tests after my changes so that we can merge this?

No problem. Done :)
(Some failing tests seem to be related to code coverage.)

**Pull request merged**